### PR TITLE
Handle error consistently on the backend using custom drf error handler

### LINKED ISSUE #623: Handle validation, errors, and business logic in the API consistently
## Problem

Currently, we have validation, error handling, and business logic spread across the `db` module, models, viewsets, and serializers to make the API work. This is confusing to work with.

## Proposed solution

- Organize the code with a clear separation of concerns.
- Document how the code is organized for reference.

## Additional context

Related to:
- #582 
- #560 
- #823

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #623

This PR introduces adds features to `Mathesar Error Handler` to capture `non ApiException` and convert it to equivalent `APIException`

Capture NonApi Exception - Used as a last resort, it captures any exceptions that do not conform to the api spec and use a default map to get the equivalent ApiException, if an equivalent ApiException does not exist, it converts it to a UnknownError(error code: 4999), with the message as the string representation of the exception that is thrown.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
Blocked by #991

File: mathesar/api/exceptions/data_import_exceptions/exceptions.py [hunk 1]
```diff
@@ -16,3 +16,36 @@ def __init__(
             status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
     ):
         super().__init__(exception, self.error_code, message, field, details, status_code)
+
+
+class URLDownloadErrorAPIException(MathesarAPIException):
+
+    def __init__(
+            self, exception, error_code=ErrorCodes.URLDownloadError.value,
```
Comment (kgodey) [inline mathesar/api/exceptions/data_import_exceptions/exceptions.py:24]: @silentninja please use one argument per line consistently, these are three arguments.

Comment (kgodey) [reply mathesar/api/exceptions/data_import_exceptions/exceptions.py:24]: also applies to other instances

File: mathesar/api/exceptions/exception_converters.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from rest_framework_friendly_errors.settings import FRIENDLY_EXCEPTION_DICT
+
+from mathesar.api.exceptions.error_codes import ErrorCodes
+
+
+def default_api_exception_converter(exc, response):
+    if isinstance(response.data, list):
+        converted_data = []
+        for data in response.data:
+            new_data = {'code': ErrorCodes.NonClassifiedError.value,
```
Comment (kgodey) [inline mathesar/api/exceptions/exception_converters.py:11]: please use consistent style, one key/value pair per line, none of the pairs should not be on the same line as curly braces. also applies to other instances.

File: mathesar/api/exceptions/exception_mappers.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from psycopg2.errors import CheckViolation, UniqueViolation
+
+from mathesar.api.exceptions.database_exceptions import (
+    exceptions as database_api_exceptions,
```
Comment (kgodey) [inline mathesar/api/exceptions/exception_mappers.py:4]: why do you need to rename these?

**Review by kgodey — APPROVED** [review]
Looks good, nice work, it's great to see the standardization coming together. Please fix formatting and merge.

File: mathesar/api/exceptions/exception_mappers.py (hunk 3 above)
Comment (silentninja) [reply mathesar/api/exceptions/exception_mappers.py:4]: We might be importing from various exception modules which are similar in name

**Pull request merged**